# Week 2 assignments 
In this week's assignments, you will practice data cleaning and validation, feature extraction, and feature engineering. You will use Pandas to write scripts to automate the data wrangling parts and [Great Expectations](https://greatexpectations.io/) for data validation.

**Guidelines for submitting assignments**:
- For each assignment, a code skeleton is provided. Please put your solutions in between the `### START CODE HERE` and `### END CODE HERE` code comments. 
- Some assignments also require you to answer questions or capture screenshots in order to earn points. Please put all your answers and the required screenshots into a single PDF file. For each answer or screenshot, please clearly indicate which assignment it corresponds to in your PDF file. 
- When preparing your submission, be sure to include your assignment notebook with code cell outputs. It's important that these outputs are current and reflect the latest state of your code, as your grades may depend on them. Additionally, please include the PDF file that contains your answers and screenshots in your submission. In the last assignment, you will need to add some code to `etl.py`. Besides this assignment notebook and the PDF file, please also include `etl.py` in your submission. 

Import all the packages needed in the assignments and set global variables.

In [1]:
import pickle
from pathlib import Path
from typing import Tuple, List, Optional

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import great_expectations as gx

from IPython.display import display

from sklearn.ensemble import RandomForestRegressor
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.preprocessing import TargetEncoder

from thefuzz import fuzz, process

from great_expectations.data_context import EphemeralDataContext, FileDataContext
from great_expectations.datasource.fluent import BatchRequest
from great_expectations.exceptions import DataContextError
from great_expectations.expectations.expectation import Expectation
from great_expectations.checkpoint.types.checkpoint_result import CheckpointResult


# Random seed for making the assignments reproducible
RANDOM_SEED = 42

# Path of current working directory
WORKING_DIR = Path.cwd()


## Assignment 1: Data extraction (2 points)
We will be working with a synthetic housing price dataset (based on [this](https://www.kaggle.com/datasets/htagholdings/property-sales) and [this](https://www.kaggle.com/datasets/harlfoxem/housesalesprediction) dataset), which lists house prices at the Canberra area in Australia from 2009 to 2016. The data description is provided in the file [Data_description.csv](./Data_description.csv). We will use this dataset to train a regression model to predict prices of houses given some info about the houses. 

### 1a) Load the data
Let's first load the training dataset, which consists of two files. The file `deals.csv` holds the price information for the sold houses along with some minor information about the houses. The file `house_info.json` has more detailed information about the houses. Complete the function `file_reader` that loads both files into Pandas DataFrames. 

Hints:
1. The `.json` file is saved in a "list-like" format, use parameter "`orient='records'`".


In [2]:
def file_reader(path: Path) -> Tuple[pd.DataFrame, pd.DataFrame]:
    """
    Read files of deals and house information into DataFrames
    Args:
        path (Path): Path to the folder where the files exist.
    Returns:
        a tuple consisting of a Pandas DataFrame of deals and a Pandas DataFrame of house information
    """
    
    price_file_path = path / "deals.csv"
    house_file_path = path / "house_info.json" 

    ### START CODE HERE
    price_df = pd.read_csv(price_file_path)
    house_df = pd.read_json(house_file_path, orient='records')
    return price_df, house_df
    ### END CODE HERE



Now, run the cell below to call the function you just created. Let's take a quick look at the DataFrames.

In [3]:
train_path = WORKING_DIR / "data" / "reference" / "train"
prices, house_data = file_reader(train_path)

print(f"DataFrame 'price' ({prices.shape[0]} rows):") 
display(prices.head())
print(f"\nDataFrame 'house_data' ({house_data.shape[0]} rows):") 
display(house_data.head())

DataFrame 'price' (11046 rows):


datesold   price  building_year  bedrooms  postcode           area
0  01-April-2009  437100           1977         4      2906         Conder
1  01-April-2009  461300           1969         4      2615  Kippax Centre
2  01-April-2009  280900           1966         3      2615        Higgins
3  01-April-2009  351600           1978         3      2620       Tinderry
4  01-April-2009  431300           2005         3      2914         Bonner


DataFrame 'house_data' (11046 rows):


date  postcode       area  bedrooms  bathrooms  yr_built  \
0 2011-10-24      2906     Gordon         4       3.00      1997   
1 2010-10-02      2605     Curtin         4       2.00      1927   
2 2015-09-25      2615  Macgregor         3       1.75      1976   
3 2011-11-03      2906      Banks         4       2.00      1977   
4 2014-12-16      2914     Bonner         3       2.50      2004   

   yr_renovated     condition  grade  floors  sqft_living  sqft_lot  \
0           NaN  Satisfactory      9     2.0         2875     13413   
1           NaN          Good      8     2.0         2450      5346   
2           NaN  Satisfactory      7     1.0         1735      8048   
3           NaN  Satisfactory      8     1.0         2555     12448   
4           NaN  Satisfactory      8     2.5         1590      1750   

   sqft_above  sqft_basement  sqft_living15  sqft_lot15  waterfront  view  \
0        2875              0         2765.0     13500.0           0     0   
1        1940            510         2200.0      5535.0           0     0   
2        1235            500         1734.0      7649.0           0     0   
3        1585            970         2341.0     10026.0           0     0   
4        1410            180         1588.0      2120.0           0     0   

   distance prev_owner  
0     20.10    Gregory  
1      7.62    Barrett  
2     13.04     Bowman  
3     21.40     Wilson  
4     14.01      Allen

Expected outputs:

<img src="./images/ex1-outputs.png" width=1200/>

It seems that both DataFrames share a date column as well as four other columns (building year, number of bedrooms, postcode and area), so we can use these columns to merge the datasets into one. To be able to do so, we need to make sure the column names and types are same in both DataFrames.

### 1b) Check column types
Let's first check the data type of the shared columns. Complete the `show_col_dtype` function to print the dtype of a given column of a given DataFrame.

In [4]:
def show_col_dtype(df: pd.DataFrame, col_name: str) -> None:
    """
    Print the dtype of a column of a DataFrame
    Args:
        df (DataFrame): The DataFrame whose column needs to be checked
        col_name(str): the name of the column to be checked
    """
    ### START CODE HERE
    col_type = df[col_name].dtype
    ### END CODE HERE

    print(f"The dtype of column {col_name} is: {col_type} ")



In [5]:
# date column
show_col_dtype(prices, "datesold")
show_col_dtype(house_data, "date")

# building year column
show_col_dtype(prices, "building_year")
show_col_dtype(house_data, "yr_built")

# bedroom number column
show_col_dtype(prices, "bedrooms")
show_col_dtype(house_data, "bedrooms")

# postcode column
show_col_dtype(prices, "postcode")
show_col_dtype(house_data, "postcode")

# area column
show_col_dtype(prices, "area")
show_col_dtype(house_data, "area")


The dtype of column datesold is: object 
The dtype of column date is: datetime64[ns] 
The dtype of column building_year is: int64 
The dtype of column yr_built is: int64 
The dtype of column bedrooms is: int64 
The dtype of column bedrooms is: int64 
The dtype of column postcode is: int64 
The dtype of column postcode is: int64 
The dtype of column area is: object 
The dtype of column area is: object 


From the output, you should see that only the date columns have different dtypes:

The dtype of column datesold is: object 

The dtype of column date is: datetime64[ns] 

### 1c) Merge price and house_data DataFrames
Now, it seems that only the column `datesold` needs to be converted to datetime format. Additionally, one of the building year columns needs to be renamed to the same as the other one. Do the following in the `dataframe_merge` function:
1. Convert the `datesold` column in the price DataFrame into `panda.datetime` format and store it in a new column named 'date' in the same DataFrame. (Do not delete the `datesold` column for now.)
1. Rename the column `building_year` to `yr_built` in the price DataFrame.
1. Merge the DataFrames using columns `date`, `postcode`, `bedrooms`, `area` and `yr_built` as keys.

The following functions may be helpful:
- [pandas.to_datetime](https://pandas.pydata.org/pandas-docs/version/1.4/reference/api/pandas.to_datetime.html)
- [pandas.DataFrame.rename](https://pandas.pydata.org/pandas-docs/version/1.4/reference/api/pandas.DataFrame.rename.html)
- [pandas.merge](https://pandas.pydata.org/pandas-docs/version/1.4/reference/api/pandas.DataFrame.merge.html)

NOTE: _In reality, we could not rely on these 5 columns to form a unique indicator for the merge to succeed. However, in the scope of this exercise we know a priori that this property holds for our data._  

In [6]:
def dataframe_merger(prices: pd.DataFrame, house_data: pd.DataFrame) -> pd.DataFrame:
    """
    Merge the two DataFrames given as inputs
    Args:
        prices (DataFrame): A pandas DataFrame holding the price information
        house_data (DataFrame): A pandas DataFrame holding the detailed information about the sold houses
    Returns:
        df: The merged pandas DataFrame
    """

    ### START CODE HERE
    prices['date'] = pd.to_datetime(prices['datesold'])
    prices = prices.rename(columns={'building_year':'yr_built'})
    return prices.merge(house_data, on=['date', 'postcode', 'bedrooms', 'area', 'yr_built'])
    ### END CODE HERE


In [7]:
train_df = dataframe_merger(prices, house_data)
print(f"DataFrame 'train_df' ({train_df.shape[0]} rows):") 
display(train_df.head())
print(f"The merged DataFrame should have {len(train_df.columns)} columns: {train_df.columns}")

DataFrame 'train_df' (11046 rows):


datesold   price  yr_built  bedrooms  postcode           area  \
0  01-April-2009  437100      1977         4      2906         Conder   
1  01-April-2009  461300      1969         4      2615  Kippax Centre   
2  01-April-2009  280900      1966         3      2615        Higgins   
3  01-April-2009  351600      1978         3      2620       Tinderry   
4  01-April-2009  431300      2005         3      2914         Bonner   

        date  bathrooms  yr_renovated     condition  ...  sqft_living  \
0 2009-04-01       2.50           NaN  satisfactory  ...         2280   
1 2009-04-01       2.25           NaN  Satisfactory  ...         2220   
2 2009-04-01       1.25           NaN  Satisfactory  ...         1130   
3 2009-04-01       2.00           NaN  Satisfactory  ...         1800   
4 2009-04-01       2.00           NaN  Satisfactory  ...         1795   

   sqft_lot  sqft_above  sqft_basement  sqft_living15  sqft_lot15  waterfront  \
0      8694        1380            900         1877.0      8052.0           0   
1     14833        1610            610         2202.0     12620.0           0   
2      8867        1130              0         1466.0      8594.0           0   
3     11431        1420            380         2032.0     11044.0           0   
4      2248        1635            160         1784.0      2714.0           0   

   view  distance  prev_owner  
0     0     20.14      Bailey  
1     0     11.74     Buckley  
2     0     10.90      Morgan  
3     0      9.48     Cumming  
4     0     13.02     Roberts  

[5 rows x 22 columns]

The merged DataFrame should have 22 columns: Index(['datesold', 'price', 'yr_built', 'bedrooms', 'postcode', 'area', 'date',
       'bathrooms', 'yr_renovated', 'condition', 'grade', 'floors',
       'sqft_living', 'sqft_lot', 'sqft_above', 'sqft_basement',
       'sqft_living15', 'sqft_lot15', 'waterfront', 'view', 'distance',
       'prev_owner'],
      dtype='object')


Expected output:

DataFrame 'train_df' (11046 rows):
| | datesold | price | yr_built | bedrooms | postcode | area | date | bathrooms | yr_renovated | condition | ... | 
| :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
| 0 | 01-April-2009 | 437100 | 1977 | 4 | 2906 | Conder | 2009-04-01 | 2.50 | NaN | satisfactory | ... | 
| 1 | 01-April-2009 | 461300 | 1969 | 4	| 2615 | Kippax Centre | 2009-04-01 | 2.25 | NaN | Satisfactory	| ... |	
| 2 | 01-April-2009 | 280900 | 1966 | 3	| 2615 | Higgins | 2009-04-01 |	1.25 | NaN | Satisfactory | ... |	
| 3 | 01-April-2009 | 351600 | 1978 | 3 | 2620 | Tinderry | 2009-04-01 | 2.00 |	NaN | Satisfactory | ... |	
| 4 | 01-April-2009 | 431300 | 2005 | 3 | 2914 | Bonner | 2009-04-01 | 2.00	| NaN | Satisfactory | ... |	

The merged DataFrame should have 22 columns: Index(['datesold', 'price', 'yr_built', 'bedrooms', 'postcode', 'area', 'date',
       'bathrooms', 'yr_renovated', 'condition', 'grade', 'floors',
       'sqft_living', 'sqft_lot', 'sqft_above', 'sqft_basement',
       'sqft_living15', 'sqft_lot15', 'waterfront', 'view', 'distance',
       'prev_owner'],
      dtype='object')

## Assignment 2: Data cleaning (2 points)
### 2a) Missing values
We have successfully extracted our data, we need to do some cleaning. By now we've noticed there are missing values in some columns. Let's start by checking which columns contain missing values.

In [8]:
print(train_df.isnull().sum())

datesold             0
price                0
yr_built             0
bedrooms             0
postcode             0
area                 0
date                 0
bathrooms            0
yr_renovated     11045
condition            0
grade                0
floors               0
sqft_living          0
sqft_lot             0
sqft_above           0
sqft_basement        0
sqft_living15      135
sqft_lot15        1079
waterfront           0
view                 0
distance             0
prev_owner           0
dtype: int64


The column `yr_renovated` contains $\frac{11045}{11046}$ missing values. With such a high fraction, it is not likely that any imputation scheme would result in a usable feature. So let's just drop that column as useless. While we're at it, we might also drop some other columns as well: 
- The name of the previous owner will not help in our prediction task, so the column `prev_owner` can be dropped as well. 
- The column `datesold` contains redundant duplicate information (the same info is included in the column `date`), so we won't need that either. 
- The values in the `sqft_living` column represent the summation of the values in the `sqft_above` and `sqft_basement` columns, so any one of the `sqft_above` and `sqft_basement` columns is individually redundant.

The columns `sqft_living15` and `sqft_lot15` have roughly only 1.5% and 10% of values missing respectively so we'll be able to use them. We will deal with the missing values in these columns later in the feature engineering part.

Remove the three columns `yr_renovated`, `prev_owner`, `datesold`, and `sqft_above` from our DataFrame.

In [9]:
def drop_futile_columns(df: pd.DataFrame) -> None:
    """
    Removes unneeded columns from the argument DataFrame
    Args:
        df (DataFrame): A pandas DataFrame holding all of the housing data
    """

    ### START CODE HERE
    df.drop(columns=['yr_renovated', 'prev_owner', 'datesold', 'sqft_above'], inplace=True)
    ### END CODE HERE


Run the cell below to see if our function works and the columns are indeed dropped. At this point, we should have 18 columns remaining.

In [10]:
drop_futile_columns(train_df)
print("The remaining columns are:", train_df.columns)

print("The total number of columns is", len(train_df.columns))

The remaining columns are: Index(['price', 'yr_built', 'bedrooms', 'postcode', 'area', 'date',
       'bathrooms', 'condition', 'grade', 'floors', 'sqft_living', 'sqft_lot',
       'sqft_basement', 'sqft_living15', 'sqft_lot15', 'waterfront', 'view',
       'distance'],
      dtype='object')
The total number of columns is 18


Expected output:

```text
The remaining columns are: Index(['price', 'yr_built', 'bedrooms', 'postcode', 'area', 'date',
       'bathrooms', 'condition', 'grade', 'floors', 'sqft_living', 'sqft_lot',
       'sqft_basement', 'sqft_living15', 'sqft_lot15', 'waterfront', 'view',
       'distance'],
      dtype='object')
The total number of columns is 18
```

### 2b) Inconsistent entries

It's very common for the extracted data to contain inconsistencies. Let's take a closer look at our columns to see if we can spot some.

In [11]:
display(train_df.describe())

price      yr_built      bedrooms      postcode     bathrooms  \
count  1.104600e+04  11046.000000  11046.000000  11046.000000  11046.000000   
mean   5.925457e+05   1972.844197      3.526073   2747.149285      2.213064   
std    2.554195e+05     27.100996      0.695975    148.665660      0.489251   
min    5.090000e+04   1900.000000      0.000000   2600.000000      1.000000   
25%    4.458000e+05   1961.000000      3.000000   2611.000000      2.000000   
50%    5.366000e+05   1977.000000      3.000000   2617.000000      2.000000   
75%    6.613000e+05   1992.000000      4.000000   2905.000000      2.500000   
max    7.989300e+06   2015.000000      5.000000   2914.000000      5.000000   

              grade        floors   sqft_living       sqft_lot  sqft_basement  \
count  11046.000000  11046.000000  11046.000000   11046.000000   11046.000000   
mean       8.171103      1.553639   2348.855694   12675.191744     270.469853   
std        1.066406      0.450539    706.753347   15545.277547     306.938218   
min        6.000000      1.000000    720.000000    1167.000000       0.000000   
25%        7.000000      1.000000   1795.000000    5837.250000      10.000000   
50%        8.000000      1.500000   2220.000000    9357.000000     130.000000   
75%        9.000000      2.000000   2780.000000   14005.750000     470.000000   
max       12.000000      3.000000   8640.000000  482350.000000    2280.000000   

       sqft_living15     sqft_lot15    waterfront          view      distance  
count   10911.000000    9967.000000  11046.000000  11046.000000  11046.000000  
mean     2299.625516   11184.507776      0.000272      0.014938    147.499479  
std       593.615570   10507.371020      0.016479      0.211600   1340.088375  
min      1208.000000    1404.000000      0.000000      0.000000      0.000000  
25%      1828.000000    5656.500000      0.000000      0.000000      8.080000  
50%      2176.000000    8905.000000      0.000000      0.000000     11.225000  
75%      2683.000000   12701.000000      0.000000      0.000000     13.990000  
max      4612.000000  182404.000000      1.000000      4.000000  22210.000000

This gives us an overview of all numerical data in the form of some statistics. In general, everything else seems to be in order, but the column `distance` seems to have some extremely high values. The data description says that this column tells the distance from the city center in km, so the max value of 22210.0 is clearly wrong. Let's take a further look at this column. Run the cell below:

In [12]:
df_low = train_df[train_df['distance']<100]
df_high = train_df[train_df['distance']>=100]
print(f"df_low ({len(df_low.index)} rows):{df_low['distance'].describe()}")
print(f"\ndf_high ({len(df_high.index)} rows):{df_high['distance'].describe()}")

df_low (10912 rows):count    10912.000000
mean        11.088641
std          4.907516
min          0.000000
25%          7.967500
50%         11.160000
75%         13.880000
max         45.680000
Name: distance, dtype: float64

df_high (134 rows):count      134.000000
mean     11255.820896
std       4825.372992
min        860.000000
25%       8377.500000
50%      10970.000000
75%      13757.500000
max      22210.000000
Name: distance, dtype: float64


It seems there are 134 rows where the distance has been stated in metres instead of kilometres. Complete the `correct_distance_unit` function that changes those values into kilometers. 

In [13]:
def correct_distance_unit(df: pd.DataFrame) -> None:
    """
    Correct the falsely input values in column 'distance'
    Args:
        df (DataFrame): A Pandas DataFrame holding all of the housing data
    """

    ### START CODE HERE
    df.loc[df['distance']>=100, 'distance'] /= 1000
    return df
    ### END CODE HERE



Let's run the cell below to see if our function works. The maximum distance should now be 45.68 km.

In [14]:
correct_distance_unit(train_df)
print(train_df['distance'].describe())

count    11046.000000
mean        11.090669
std          4.906347
min          0.000000
25%          7.990000
50%         11.160000
75%         13.877500
max         45.680000
Name: distance, dtype: float64


Expected output:

```text
count    11046.000000
mean        11.090669
std          4.906347
min          0.000000
25%          7.990000
50%         11.160000
75%         13.877500
max         45.680000
Name: distance, dtype: float64
```

### 2c) Fixing typos

Now we should have our numerical columns in order. Let's next take a look at the string columns.

In [15]:
condition_unique = train_df['condition'].unique()
print(f"Column 'condition' has {len(condition_unique)} unique values:\n{condition_unique}")
area_unique = train_df['area'].unique()
print(f"\nColumn 'area' has {len(area_unique)} unique values:\n{area_unique}")

Column 'condition' has 16 unique values:
['satisfactory' 'Satisfactory' 'Good' 'Excellent' 'good' 'Satisfactoryy'
 'Goood' 'Satisfactry' 'satisfatory' 'Satisfattory' 'God' 'Good '
 'excellent' 'Satisfatory' 'god' 'satisfattory']

Column 'area' has 131 unique values:
['Conder' 'Kippax Centre' 'Higgins' 'Tinderry' 'Bonner' 'Fraser' 'Kambah'
 'Duffy' 'Page' 'Kaleen' 'Kingston' 'Gowrie' 'Administrative' 'Gordon'
 'Kippax' 'Ginninderra Village' 'Bonython' 'Curtin' 'Uriarra' 'Forde'
 'Monash' 'Waramanga' 'Scullin' 'Pearce' 'Fairbairn Raaf' 'Amaroo'
 'Nicholls' 'Macarthur' 'Dickson' 'Theodore' 'Lyneham' "O'connor"
 'Hackett' 'Downer' 'Ainslie' 'Acton' 'Stirling' 'Belconnen Dc' 'Watson'
 'Belconnen' 'Melba' 'Causeway' 'Isabella Plains' 'Holder' 'Russell'
 'Spence' 'Red Hill' 'Holt' 'Wanniassa' 'Hughes' 'Lawson' 'Cook' 'Torrens'
 'Chapman' 'Harman' 'Dunlop' 'Farrer' 'Jamison Centre' 'Hawker'
 'Charnwood' 'Weston Creek' 'Mawson' 'Turner' 'Latham' 'Oxley' 'Ngunnawal'
 'Palmerston' 'Florey' 'Pierc

The column `area` seems to be ok, but the column `condition` contains several typos. 

Start by lowercasing the conditions and removing any unnecessary trailing white space.

In [16]:
def string_transformer(df: pd.DataFrame) -> None:
    """
    Lowercases all values in the column 'condition' and removes trailing white space.
    Args:
        df (DataFrame): A pandas DataFrame holding all of the housing data
    """

    ### START CODE HERE
    df['condition'] = df['condition'].str.lower()
    df['condition'] = df['condition'].str.strip()
    return df
    ### END CODE HERE



Let's apply this function to our training data to see what changed. Run the cell below:

In [17]:
string_transformer(train_df)
print("The column 'condition' contains following unique values:\n", train_df['condition'].unique())

The column 'condition' contains following unique values:
 ['satisfactory' 'good' 'excellent' 'satisfactoryy' 'goood' 'satisfactry'
 'satisfatory' 'satisfattory' 'god']


Expected output:
```text
The column 'condition' contains following unique values:
 ['satisfactory' 'good' 'excellent' 'satisfactoryy' 'goood' 'satisfactry'
 'satisfatory' 'satisfattory' 'god']
```

We see that the data contains three conditions, namely: 'excellent', 'good' and 'satisfactory'. Other values are clearly typos. Let's use [TheFuzz](https://github.com/seatgeek/thefuzz) library (already imported above) to automate the process of correcting typos. 

Implement the following to complete the `typo_fixer` function:
1. Loop over the condition column and find the best approximate match from a list of correct conditions for each entry in the condition column. You can use [process.extractOne](https://github.com/seatgeek/thefuzz#process) to find the single best match for an entry and use `fuzz.ratio` as the scorer inside the `extractOne` function.
2. Update the column `condition` by replacing each entry with its best approximate match if the similarity score is greater than the specified threshold. Otherwise, keep the entry unchanged. In this way, we fix only those strings that feel confident enough to be typos of the correct conditions. Using a threshold is a good practice so that we are not too aggressive in correcting the typos. 
3. Create a new column `similarity_scores`, where you store the similarity score of each (original entry)-(approximate match) pair.

In [18]:
def typo_fixer(df: pd.DataFrame, threshold: float, correct_condition_values: List[str]) -> None:
    """
    Uses fuzzy string matching to fix typos in the column 'condition'. It loops through each entry in the column and 
    replaces them with suggested corrections if the similarity score is high enough. 
    Args:
        df (DataFrame): A pandas DataFrame holding all of the housing data
        threshold (int): A number between 0-100. Only the entries with score above this number are replaced.
        correct_condition_values (List): A list of correct strings that we hope the condition column to include. For example, correct_ones=['excellent', 'good', 'satisfactory'] in the case of the training dataset.
    """

    ### START CODE HERE
    scores = []
    for idx, row in df.iterrows():
        best_match, score = process.extractOne(row['condition'], correct_condition_values, scorer=fuzz.ratio)
        if score > threshold:
            df['condition'][idx] = best_match
        scores.append(score)
    df['similarity_scores'] = scores
    return df
    ### END CODE HERE


In [19]:
threshold = 80
typo_fixer(train_df, threshold, ['excellent', 'good', 'satisfactory'])
print("The column 'condition' contains following unique values:\n", train_df['condition'].unique())
print(f"The lowest similarity score is {train_df['similarity_scores'].min()}")

The column 'condition' contains following unique values:
 ['satisfactory' 'good' 'excellent']
The lowest similarity score is 86


You're expected to see all typos are corrected. Expected output:
```text
The column 'condition' contains following unique values:
 ['satisfactory' 'good' 'excellent']
The lowest similarity score is 86
```

Now that we have our data extraction and cleaning operations in place, let's conclude the first two assignments by wrapping everything up to this point inside a single function `data_extraction`. It loads the price and house info datasets, merges them into a single DataFrame and polishes the DataFrame.

Run the cell below to define the data extraction function.

In [20]:
def data_extraction(path: Path, correct_condition_values: List[str]) -> pd.DataFrame:
    """
    The entire data extraction/cleaning pipeline wrapped inside a single function.
    Args:
        path (Path): Path to the folder where the files exist.
        correct_condition_values (List): A list of correct strings that we hope the condition column to include.
    Returns:
        df (DataFrame): A pandas DataFrame holding all of the (cleaned) housing data.
    """
    threshold = 80
    prices, house_data = file_reader(path)
    df = dataframe_merger(prices, house_data)
    drop_futile_columns(df)
    correct_distance_unit(df)
    string_transformer(df)
    typo_fixer(df, threshold, correct_condition_values)
    return df



In [21]:
# Create a training DataFrame that goes through the data extraction and cleaning processes. This DataFrame will be used in the following assignments
train_path = WORKING_DIR / "data" / "reference" / "train"
train_df = data_extraction(train_path, ['excellent', 'good', 'satisfactory'])
train_df.head(5)

price  yr_built  bedrooms  postcode           area       date  bathrooms  \
0  437100      1977         4      2906         Conder 2009-04-01       2.50   
1  461300      1969         4      2615  Kippax Centre 2009-04-01       2.25   
2  280900      1966         3      2615        Higgins 2009-04-01       1.25   
3  351600      1978         3      2620       Tinderry 2009-04-01       2.00   
4  431300      2005         3      2914         Bonner 2009-04-01       2.00   

      condition  grade  floors  sqft_living  sqft_lot  sqft_basement  \
0  satisfactory      7     1.0         2280      8694            900   
1  satisfactory      8     1.0         2220     14833            610   
2  satisfactory      7     1.0         1130      8867              0   
3  satisfactory      8     1.0         1800     11431            380   
4  satisfactory      8     2.0         1795      2248            160   

   sqft_living15  sqft_lot15  waterfront  view  distance  similarity_scores  
0         1877.0      8052.0           0     0     20.14                100  
1         2202.0     12620.0           0     0     11.74                100  
2         1466.0      8594.0           0     0     10.90                100  
3         2032.0     11044.0           0     0      9.48                100  
4         1784.0      2714.0           0     0     13.02                100

Expected output:

<img src="./images/extracted_train_df.png" width=1300/>

## Assignment 3: Data validation with Great Expectations (2 points)

In this assignment, we practice the use of Great Expectations as a tool for data validation. We provide you with some code to help you instantiate and get started with Great Expectations. Running the code cell below creates a folder `gx` in your working directory with all things related to working with Great Expectations. 

In [22]:
housing_datasource_name = "housing_datasource"
housing_expectation_suite_name = "housing_expectation_suite"

# Instantiate a Data Context and save it in filesystem.
context = gx.get_context(project_root_dir=str(WORKING_DIR))

# Connect to data in our housing DataFrame.
try:
    # Create a new Data Source in the Data Context
    datasource = context.sources.add_pandas(name=housing_datasource_name)
except DataContextError:
    # The Data Source already exists in the Data Context
    datasource = context.get_datasource(housing_datasource_name)

try:
    # Create a new DataFrame Data Asset
    training_data_asset = datasource.add_dataframe_asset(name="training_data")
except ValueError:
    # The Data Asset already exists
    training_data_asset = datasource.get_asset("training_data")

# Request all data in the DataFrame as a single batch
training_batch_request = training_data_asset.build_batch_request(dataframe=train_df)

# Create an Expectation Suite 
context.add_or_update_expectation_suite(housing_expectation_suite_name)

# Create a Validator
housing_validator = context.get_validator(
    batch_request=training_batch_request,
    expectation_suite_name=housing_expectation_suite_name
)
display(housing_validator.head())

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

price  yr_built  bedrooms  postcode           area       date  bathrooms  \
0  437100      1977         4      2906         Conder 2009-04-01       2.50   
1  461300      1969         4      2615  Kippax Centre 2009-04-01       2.25   
2  280900      1966         3      2615        Higgins 2009-04-01       1.25   
3  351600      1978         3      2620       Tinderry 2009-04-01       2.00   
4  431300      2005         3      2914         Bonner 2009-04-01       2.00   

      condition  grade  floors  sqft_living  sqft_lot  sqft_basement  \
0  satisfactory      7     1.0         2280      8694            900   
1  satisfactory      8     1.0         2220     14833            610   
2  satisfactory      7     1.0         1130      8867              0   
3  satisfactory      8     1.0         1800     11431            380   
4  satisfactory      8     2.0         1795      2248            160   

   sqft_living15  sqft_lot15  waterfront  view  distance  similarity_scores  
0         1877.0      8052.0           0     0     20.14                100  
1         2202.0     12620.0           0     0     11.74                100  
2         1466.0      8594.0           0     0     10.90                100  
3         2032.0     11044.0           0     0      9.48                100  
4         1784.0      2714.0           0     0     13.02                100

### 3a) Create Expectations
Your task is to create the following Expectations using the `expect_*` methods provided by the Validator (`housing_validator`). Additionally, recall that the `expect_*` methods will also validate the data loaded into the Validator, so please also store the validation results in a list.
1. DataFrame `df` should contain all the columns in the same order as `train_df` (the training DataFrame produced by the `data_extraction` function).
1. Column `price` can not contain NaN values.
1. Column `yr_built` shouldn't have values smaller than 1917.
1. Column `bedrooms` should be an integer.
1. Column `condition` should contain only values from the set {'satisfactory', 'good', 'excellent'}.

Hints: The following validation functions may be helpful. You can find more information of their usage [here](https://greatexpectations.io/expectations/). 
- expect_table_columns_to_match_ordered_list
- expect_column_values_to_not_be_null
- expect_column_values_to_be_between
- expect_column_values_to_be_of_type
- expect_column_values_to_be_in_set

In [23]:
test_results = []

### START CODE HERE
# e.g., test_results.append(...)
test_results.append(housing_validator.expect_table_columns_to_match_ordered_list(train_df.columns.values.tolist()))
test_results.append(housing_validator.expect_column_values_to_not_be_Null('price'))
test_results.append(housing_validator.expect_column_values_to_be_between('yr_built', min_value=1917))
test_results.append(housing_validator.expect_column_values_to_be_of_type('bedrooms', 'int'))
test_results.append(housing_validator.expect_column_values_to_be_in_set('condition', {'satisfactory', 'good', 'excellent'}))
### END CODE HERE

# Print the validation results of the batch loaded into the Validator
for i, result in enumerate(test_results):
    if result['success'] is True:      
        print(f"Test {i} succeeded ({result['expectation_config']['expectation_type']})")
    else:
        print(f"Test {i} failed ({result['expectation_config']['expectation_type']})")
        for item in result['result'].items():
            print("\t", item)

Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/6 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Test 0 succeeded (expect_table_columns_to_match_ordered_list)
Test 1 succeeded (expect_column_values_to_not_be_null)
Test 2 failed (expect_column_values_to_be_between)
	 ('element_count', 11046)
	 ('unexpected_count', 437)
	 ('unexpected_percent', 3.9561832337497735)
	 ('partial_unexpected_list', [1900, 1910, 1913, 1910, 1906, 1915, 1906, 1910, 1906, 1912, 1905, 1909, 1909, 1906, 1909, 1910, 1903, 1906, 1910, 1916])
	 ('missing_count', 0)
	 ('missing_percent', 0.0)
	 ('unexpected_percent_total', 3.9561832337497735)
	 ('unexpected_percent_nonmissing', 3.9561832337497735)
Test 3 succeeded (expect_column_values_to_be_of_type)
Test 4 succeeded (expect_column_values_to_be_in_set)


Expected output:
```text
Test 0 succeeded (expect_table_columns_to_match_ordered_list)
Test 1 succeeded (expect_column_values_to_not_be_null)
Test 2 failed (expect_column_values_to_be_between)
	 ('element_count', 11046)
	 ('unexpected_count', 437)
	 ('unexpected_percent', 3.9561832337497735)
	 ('partial_unexpected_list', [1900, 1910, 1913, 1910, 1906, 1915, 1906, 1910, 1906, 1912, 1905, 1909, 1909, 1906, 1909, 1910, 1903, 1906, 1910, 1916])
	 ('missing_count', 0)
	 ('missing_percent', 0.0)
	 ('unexpected_percent_total', 3.9561832337497735)
	 ('unexpected_percent_nonmissing', 3.9561832337497735)
Test 3 succeeded (expect_column_values_to_be_of_type)
Test 4 succeeded (expect_column_values_to_be_in_set)
```

We see that the Expectation regarding the building year is not met. There are $\frac{437}{11046}=3.96\%$ of entries not conforming to our Expectation. There are (at least) three ways in which we can deal with this: 
1. We can manually check the min value of the column `yr_built` (in fact, we already have the info in Assignment 2b) and use that as the lower bound. 
1. We can use the keyword argument `mostly` to specify a fraction of the data that must conform with our expectations. This allows for some values to be outside of the defined interval. 
1. We can try to define the boundaries automatically (see below). 

Which ever approach we take, we can just redefine the Expectation and that will overwrite the old one.

Some (but not all) Expectations allow for automatic inferring of boundaries for values using the batch loaded into the Validator. This is called auto-initializing. You can check if an Expectation can be auto-initialized as follows.

In [24]:
boolean_1 = Expectation.is_expectation_auto_initializing(name="expect_table_columns_to_match_ordered_list")
boolean_2 = Expectation.is_expectation_auto_initializing(name="expect_column_values_to_be_between")
print(f"\nboolean_1 = {boolean_1} and boolean_2 = {boolean_2}")

The Expectation expect_table_columns_to_match_ordered_list is not able to be auto-initialized.
The Expectation expect_column_values_to_be_between is able to be auto-initialized. Please run by using the auto=True parameter.

boolean_1 = False and boolean_2 = True


Using auto-initializing is convenient especially when working with a large number of covariates. However, it requires that we have access to some data, which has already been validated (luckily, we do). Also, auto-initialization might produce some rules that are too strict or too loose, so always be careful when using them and check the results.

Update our `housing_validator` with the following Expectations: 
- The values of columns `yr_built`, `bedrooms`, and `date` should be between a minimum value and a maximum value. These values are automatically inferred by Great Expectations.
- The values of columns `area` should be in a set. The set is also automatically inferred by Great Expectations.

Similar to what you did previously, also append the resulted ExpectationValidationResults into a list. 

Hint: Check [here](https://docs.greatexpectations.io/docs/guides/expectations/how_to_use_auto_initializing_expectations/) on how to enable auto-initialization of an Expectation. 

In [25]:
test_results = []

### START CODE HERE
test_results.append(housing_validator.expect_column_values_to_be_between('yr_built', auto=True))
test_results.append(housing_validator.expect_column_values_to_be_between('bedrooms', auto=True))
test_results.append(housing_validator.expect_column_values_to_be_between('date', auto=True))
test_results.append(housing_validator.expect_column_values_to_be_in_set('area', auto=True))
### END CODE HERE

# Print the Expectation configuration that Great Expectations automatically inferred
for i, result in enumerate(test_results):
    print(f"Test {i} ({result['expectation_config']['expectation_type']}) resulted in config:")
    for item in result['expectation_config']['kwargs'].items():
        print("\t", item)

Generating Expectations:   0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

Profiling Dataset:         0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Generating Expectations:   0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

Profiling Dataset:         0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Generating Expectations:   0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

Profiling Dataset:         0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Generating Expectations:   0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/2 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

Profiling Dataset:         0%|          | 0/1 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Test 0 (expect_column_values_to_be_between) resulted in config:
	 ('column', 'yr_built')
	 ('min_value', 1900)
	 ('max_value', 2015)
	 ('mostly', 1.0)
	 ('strict_min', False)
	 ('strict_max', False)
	 ('auto', True)
	 ('batch_id', 'housing_datasource-training_data')
Test 1 (expect_column_values_to_be_between) resulted in config:
	 ('min_value', 0)
	 ('strict_min', False)
	 ('column', 'bedrooms')
	 ('strict_max', False)
	 ('max_value', 5)
	 ('mostly', 1.0)
	 ('auto', True)
	 ('batch_id', 'housing_datasource-training_data')
Test 2 (expect_column_values_to_be_between) resulted in config:
	 ('min_value', '2007-07-02T00:00:00')
	 ('strict_min', False)
	 ('column', 'date')
	 ('strict_max', False)
	 ('max_value', '2015-12-24T00:00:00')
	 ('mostly', 1.0)
	 ('auto', True)
	 ('batch_id', 'housing_datasource-training_data')
Test 3 (expect_column_values_to_be_in_set) resulted in config:
	 ('column', 'area')
	 ('value_set', ['Macquarie', 'Acton', 'Oxley', 'Red Hill', 'Evatt', 'Uriarra Forest', 'Isa

Expected output:

```text
Test 0 (expect_column_values_to_be_between) resulted in config:
	 ('column', 'yr_built')
	 ('min_value', 1900)
	 ('max_value', 2015)
	 ('mostly', 1.0)
	 ('strict_min', False)
	 ('strict_max', False)
	 ('auto', True)
	 ('batch_id', 'housing_datasource-training_data')
Test 1 (expect_column_values_to_be_between) resulted in config:
	 ('max_value', 5)
	 ('column', 'bedrooms')
	 ('min_value', 0)
	 ('strict_max', False)
	 ('mostly', 1.0)
	 ('strict_min', False)
	 ('auto', True)
	 ('batch_id', 'housing_datasource-training_data')
Test 2 (expect_column_values_to_be_between) resulted in config:
	 ('max_value', '2015-12-24T00:00:00')
	 ('column', 'date')
	 ('min_value', '2007-07-02T00:00:00')
	 ('strict_max', False)
	 ('mostly', 1.0)
	 ('strict_min', False)
	 ('auto', True)
	 ('batch_id', 'housing_datasource-training_data')
Test 3 (expect_column_values_to_be_in_set) resulted in config:
	 ('column', 'area')
	 ('value_set', ['Jamison Centre', 'Gundaroo', 'Wanniassa', 'Kaleen', ...])
	 ('mostly', 1.0)
	 ('auto', True)
	 ('batch_id', 'housing_datasource-training_data')
```

We could specify more Expectations, but the Expectations we've defined suffice for now. Recall that the Expectation Suite we have configured this far exists only in memory and has to be persisted for future use. Let's save the Expectation Suite into our Data Context.
 

In [26]:
# Persist the Expectation Suite
housing_validator.save_expectation_suite()

### 3b) Validate new data against the defined Expectations

Let's run our newly defined Expectation Suite against some test data. This is done by creating and configuring a [Checkpoint](https://docs.greatexpectations.io/docs/terms/checkpoint). Before we do that, we'll have to create a Batch Request for the new data. We have some unprocessed test data residing in a test data folder. 

Create the following functions:
1. `batch_creator`, which 
    1. creates a new DataFrame DataAsset named "test_data" to a Data Source or gets the asset from the Data Source if it's already existing, then 
    1. creates a new Batch Request for the given DataFrame and returns the new Batch Request.
1. `create_checkpoint`, which 
    1. creates a new Checkpoint with the following configurations:
        - The name of the Checkpoint should be the given `checkpoint_name`,
        - It should validate the data of the given `batch_request` using an Expectation Suite whose name is the given `expectation_suite_name`
    1. uses the created Checkpoint to run a validation, the validation running name should be the given `run_name`,
    1. finally, returns the validation results. 

Hints: 
- The following functions may be useful for implementing the `batch_creator` function: 
    - [add_dataframe_asset and build_batch_request](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/fluent/in_memory/connect_in_memory_data/#add-a-data-asset-to-the-data-source)
    - [get_asset](https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/fluent/batch_requests/how_to_request_data_from_a_data_asset#retrieve-your-data-asset)

- The following functions may be useful for implementing the `create_checkpoint` function: 
    - [add_or_update_checkpoint](https://docs.greatexpectations.io/docs/reference/api/data_context/AbstractDataContext_class#great_expectations.data_context.AbstractDataContext.add_or_update_checkpoint)
    - [(Checkpoint).run](https://docs.greatexpectations.io/docs/reference/api/checkpoint/Checkpoint_class#great_expectations.checkpoint.Checkpoint.run)

In [27]:
def batch_creator(df: pd.DataFrame, context: FileDataContext, data_source_name: str) -> BatchRequest:
    """
    Creates a new Batch Request using the given DataFrame
    Args:
        df (DataFrame): A pandas DataFrame holding the cleaned housing data. 
        context (GX FileDataContext): The current active GX Data Context 
        data_source_name (str): Name of the GX Data Source, to which the DataFrame is added
    Returns:
        new_batch_request (GX BatchRequest): The GX batch request created using df
    """
    datasource = context.get_datasource(data_source_name)
    test_asset_name = "test_data"
    
    ### START CODE HERE
    try:
        data_asset = datasource.add_dataframe_asset(name=test_asset_name)
    except ValueError:
        data_asset = datasource.get_asset(test_asset_name)
    return data_asset.build_batch_request(dataframe=df)
    ### END CODE HERE


def create_checkpoint(context: FileDataContext, batch_request: BatchRequest, checkpoint_name: str, expectation_suite_name: str, run_name: str) -> CheckpointResult:
    """
    Creates a new GX Checkpoint from the argument batch_request
    Args:
        context (GX FileDataContext): The current active context 
        new_batch_request (GX BatchRequest): A GX batch request used to create the Checkpoint
        checkpoint_name (str): Name of the Checkpoint
        expectation_suite_name (str): Name of the Expectation Suite used to validate the data
        run_name (str): Name of the validation running
    Returns:
        checkpoint_result (GX CheckpointResult): 
    """
    
    ### START CODE HERE
    ckpt = context.add_or_update_checkpoint(name=checkpoint_name, batch_request=batch_request, expectation_suite_name=expectation_suite_name)
    return ckpt.run(run_name=run_name)
    ### END CODE HERE



Run the cell below to call both functions. This should create the Checkpoint and returns a Checkpoint result.

In [28]:
# Test data location
test_path = WORKING_DIR / "data" / "reference" / "test"

# Create the batch request and checkpoint
context = gx.get_context()
test_df = data_extraction(test_path, ['satisfactory', 'good', 'excellent'])
test_batch_request = batch_creator(
    df=test_df,
    context=context,
    data_source_name=housing_datasource_name)

test_data_checkpoint_result = create_checkpoint(context=context,
                                                batch_request=test_batch_request,
                                                checkpoint_name="test_data_checkpoint",
                                                expectation_suite_name=housing_expectation_suite_name,
                                                run_name="test_run")

Calculating Metrics:   0%|          | 0/43 [00:00<?, ?it/s]

In [29]:
context.view_validation_result(test_data_checkpoint_result)

If everything went smoothly, a page showing the result should be opened in your browser. We see that one out of eight expectations has failed. By scrolling down, we see that the failed expectation is the one which specifies that the values of column `cond_typos_fixed` has to be one of {'satisfactory', 'good', 'excellent'}. The test data seems to have new values 'tolerable' and 'poor'. You can find the result in `gx/uncommitted/data_docs/local_site/index.html`

NOTE: _If we hadn't specified a threshold when we did fuzzy string matching, the algorithm would have changed the values 'tolerable' and 'poor' to their nearest matches from {'satisfactory', 'good', 'excellent'}, which would have been very undesirable._

Do the following:
1. Update the Expectation for column `condition` to include values 'tolerable' and 'poor' as well.
1. Save the updated Expectation Suite into our Data Context.
1. Use the `create_checkpoint` function you just made to create a new Checkpoint with arguments (`test_batch_request`, `"second_test_checkpoint"`, `"housing_expectation_suite"`, `"second_test_run"`). Store the results into a variable named `second_test_checkpoint_result`.

In [30]:
### START CODE HERE
housing_validator.expect_column_values_to_be_in_set('condition', {'satisfactory', 'good', 'excellent', 'tolerable', 'poor'})
housing_validator.save_expectation_suite()

second_test_data_checkpoint_result = create_checkpoint(context=context,
                                                batch_request=test_batch_request,
                                                checkpoint_name="second_test_checkpoint",
                                                expectation_suite_name="housing_expectation_suite",
                                                run_name="second_test_run")
### END CODE HERE


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/43 [00:00<?, ?it/s]

In [31]:
context.view_validation_result(second_test_data_checkpoint_result)

Now we should see that all Expectations have been met.

NOTE: _Creating Checkpoints and validating data against a set of Expectations doesn't do anything in itself (other than create the Data Docs). We can of course manually check the Data Docs after validating each new batch of data, but it would be nice if could automate at least some part of this process. This is what the [Great Expectations Actions](https://docs.greatexpectations.io/docs/guides/validation/validation_actions/actions_lp) are for. However, we won't utilize this functionality in this exercise._

### Screenshots to be submitted for Assignment 3
1. A screenshot of the failed Checkpoint result (i.e., the Checkpoint run named "test-run").
1. A screenshot of the succeeded Checkpoint result (i.e., the Checkpoint run named "second-test-run"). 

**Note**: It's enough to show the overview and the failed/corrected Validation. 

<details>
    <summary>Example</summary>
    <figure>
        <img src="./images/test-run-overview.png" width=600/>
        <img src="./images/test-run-condition.png" width=800/>
        <figcaption>test-run</figcaption>
    </figure>
    <br />
    <figure>
        <img src="./images/second-test-run-overview.png" width=700/>
        <img src="./images/second-test-run-condition.png" width=800/>
        <figcaption>second-test-run</figcaption>
    </figure>
</details>

## Assignment 4: Feature engineering (2 points)

In this part we will touch on the subject of feature engineering. Since this subject is dealt with in more detail in the course '[Introduction to Machine Learning](https://studies.helsinki.fi/courses/course-unit/hy-CU-118207827-2021-08-01)', we try to keep it light whilst also (perhaps) covering some corners not talked about on that course. Some of the methods presented in this exercise require us to divide our data into features and targets, so let's begin by building a function to achieve that (the function is ready so you don't need to implement anything). Run the cell below:

In [32]:
def separate_X_and_y(df: pd.DataFrame, target: str) -> Tuple[pd.DataFrame, pd.Series]:
    """
    Separates the features and targets.
    Args:
        df (DataFrame): A pandas DataFrame holding the cleaned housing data
        target (str): Name of the target column
    Returns:
        X (DataFrame): A Pandas DataFrame holding the cleaned housing data without the target column
        y (Series): A pandas Series with the target values
    """
    df = df.copy()  # By copying the DataFrame we protect the original DataFrame in case we make errors in the following assignments. This way, we can always restart from this step.
    y = df[target].astype('float64')
    X = df.loc[:, df.columns != target]    
    return (X, y)



Then, we'll use this function to separate the target column from the rest for both the training and testing datasets. Let's create the entire datasets from the beginning. Again, run the cell below:

In [33]:
train_path = WORKING_DIR / "data" / "reference" / "train"
test_path = WORKING_DIR / "data" / "reference" / "test"

train_df = data_extraction(train_path, ['satisfactory', 'good', 'excellent'])
test_df = data_extraction(test_path, ['poor', 'tolerable', 'satisfactory', 'good', 'excellent'])

X_train, y_train = separate_X_and_y(train_df, target='price')
X_test, y_test = separate_X_and_y(test_df, target='price')

print(f"X_train shape: {X_train.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_test shape: {y_test.shape}")
display(X_train.head())
display(y_train.head())

X_train shape: (11046, 18)
y_train shape: (11046,)
X_test shape: (2762, 18)
y_test shape: (2762,)


yr_built  bedrooms  postcode           area       date  bathrooms  \
0      1977         4      2906         Conder 2009-04-01       2.50   
1      1969         4      2615  Kippax Centre 2009-04-01       2.25   
2      1966         3      2615        Higgins 2009-04-01       1.25   
3      1978         3      2620       Tinderry 2009-04-01       2.00   
4      2005         3      2914         Bonner 2009-04-01       2.00   

      condition  grade  floors  sqft_living  sqft_lot  sqft_basement  \
0  satisfactory      7     1.0         2280      8694            900   
1  satisfactory      8     1.0         2220     14833            610   
2  satisfactory      7     1.0         1130      8867              0   
3  satisfactory      8     1.0         1800     11431            380   
4  satisfactory      8     2.0         1795      2248            160   

   sqft_living15  sqft_lot15  waterfront  view  distance  similarity_scores  
0         1877.0      8052.0           0     0     20.14                100  
1         2202.0     12620.0           0     0     11.74                100  
2         1466.0      8594.0           0     0     10.90                100  
3         2032.0     11044.0           0     0      9.48                100  
4         1784.0      2714.0           0     0     13.02                100

0    437100.0
1    461300.0
2    280900.0
3    351600.0
4    431300.0
Name: price, dtype: float64

Expected output:
```text
X_train shape: (11046, 18)
y_train shape: (11046,)
X_test shape: (2762, 18)
y_test shape: (2762,)
```
<img src="./images/data-split-output.png" />

### 4a) Imputing missing values

As we saw in Assignment 2a), the columns `sqft_living15` and `sqft_lot15` have roughly 1.5% and 10% of values missing respectively in the training data so we could try to impute something in place of the missing values. We could use a simple imputation scheme, such as imputing either zero, mean, median or mode to both columns in place of missing values. However, we could also use a more sophisticated scheme, such as [MICE](https://www.machinelearningplus.com/machine-learning/mice-imputation/), where we use a regressor model to infer the missing values by comparing other features iteratively. We will use the [scikit-learn implementation](https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html) in this exercise. 

Your task is to complete the `impute_missing` function that uses `IterativeImputer` to impute missing values of Dataframes `X_train` and `X_test`. We use parameter `add_indicator=True` to include two new columns that indicate which values were imputed. These indicator variables will have a value of 1 where the original data was missing and 0 where it was not. This information might be valuable in model training (and inference). 

You can use the `fit_transform` method to fit the imputer on a DataFrame X and return the transformed X. Note that the fit_transform method will return a Numpy array instead of a DataFrame, and the shape of the returned Numpy array will not match the shape of the DataFrame given as an argument, so you'll have to figure a way around this.

Hint: You can use the method [get_feature_names_out](https://scikit-learn.org/stable/modules/generated/sklearn.impute.IterativeImputer.html#sklearn.impute.IterativeImputer.get_feature_names_out) to list the column names of the `IterativeImputer` object once it has been fitted.

NOTE: _The computational complexity increases very rapidly with regards to the number of features in the dataset._

In [38]:
def impute_missing(df: pd.DataFrame) -> None:
    """
    Imputes missing numerical values using MICE.
    Args:
        df (DataFrame): A pandas DataFrame holding the features
    """
    imp = IterativeImputer(random_state=RANDOM_SEED, add_indicator=True) 

    # A list of column names where missing values need to be imputed.
    # We'll not include categorical or datetime variables in the calculations
    included_columns = [True if x not in ['postcode', 'area', 'date', 'condition'] else False for x in df.columns]

    ### START CODE HERE
    cols_to_impute = df.loc[:,included_columns]
    imputed_matrix = imp.fit_transform(cols_to_impute)
    df[imp.get_feature_names_out()] = pd.DataFrame(data=imputed_matrix, columns=imp.get_feature_names_out())
    ### END CODE HERE


Let's again test the function we just created. Run the cell below:

In [39]:
impute_missing(X_train)
impute_missing(X_test)
nan_train = X_train.isnull().sum().sum()
nan_test = X_test.isnull().sum().sum()
print(f"The numbers of missing values are {nan_train} for the train set and {nan_test} for the test set. Both should be zero.")
print(f"There should be {X_train.shape[1]} columns: {X_train.columns}")

The numbers of missing values are 0 for the train set and 0 for the test set. Both should be zero.
There should be 20 columns: Index(['yr_built', 'bedrooms', 'postcode', 'area', 'date', 'bathrooms',
       'condition', 'grade', 'floors', 'sqft_living', 'sqft_lot',
       'sqft_basement', 'sqft_living15', 'sqft_lot15', 'waterfront', 'view',
       'distance', 'similarity_scores', 'missingindicator_sqft_living15',
       'missingindicator_sqft_lot15'],
      dtype='object')


Expected output: 

```text
The numbers of missing values are 0 for the train set and 0 for the test set. Both should be zero.
There should be 20 columns: Index(['yr_built', 'bedrooms', 'postcode', 'area', 'date', 'bathrooms',
       'condition', 'grade', 'floors', 'sqft_living', 'sqft_lot',
       'sqft_basement', 'sqft_living15', 'sqft_lot15', 'waterfront', 'view',
       'distance', 'similarity_scores', 'missingindicator_sqft_living15',
       'missingindicator_sqft_lot15'],
      dtype='object')

```

### 4b) Decomposing datetime and categorical variables

The column `date` is actually hoarding a bunch of potential features. Since we converted this column to `DateTime` format, we can use the attributes of this format to decompose the column into several features, which might benefit model training. 
Create a function, which takes a DataFrame and a column name for a DateTime column as arguments and
1. uses `pandas.Series.dt.*` attributes to decompose the DateTime column into three new columns: `year`, `quarter`, and `weekday`. The column `quarter` should have values 1-4 and the column `weekday` should have values 0-6, where 0 is Monday, 1 is Tuesday, and so on. (You might take a look at [this blog](https://datagy.io/pandas-datetime/)).
1. removes the original DateTime column from the DataFrame.

In [44]:
def datetime_decomposer(df: pd.DataFrame, dt_column_name: str) -> None:
    """
    Decomposes datetime values into year, quarter, and weekday.
    Args:
        df (DataFrame): A pandas DataFrame holding the features
        dt_column_name(str): The name of the datetime column
    """

    ### START CODE HERE
    dates = df[dt_column_name]
    df['year'] = dates.dt.year
    df['quarter'] = dates.dt.quarter
    df['weekday'] = dates.dt.dayofweek
    df.drop(dt_column_name, axis=1, inplace=True)
    ### END CODE HERE



Apply this function for both `X_train` and `X_test` by running the cell below. There should be 22 columns at this point.

In [45]:
datetime_decomposer(X_train, 'date')
datetime_decomposer(X_test, 'date')

print(f"There are a total of {len(X_train.columns)} columns remaining:")
print(X_train.columns)

There are a total of 22 columns remaining:
Index(['yr_built', 'bedrooms', 'postcode', 'area', 'bathrooms', 'condition',
       'grade', 'floors', 'sqft_living', 'sqft_lot', 'sqft_basement',
       'sqft_living15', 'sqft_lot15', 'waterfront', 'view', 'distance',
       'similarity_scores', 'missingindicator_sqft_living15',
       'missingindicator_sqft_lot15', 'year', 'quarter', 'weekday'],
      dtype='object')


Expected output:
```text
There are a total of 22 columns remaining:
Index(['yr_built', 'bedrooms', 'postcode', 'area', 'bathrooms', 'condition',
       'grade', 'floors', 'sqft_living', 'sqft_lot', 'sqft_basement',
       'sqft_living15', 'sqft_lot15', 'waterfront', 'view', 'distance',
       'similarity_scores', 'missingindicator_sqft_living15',
       'missingindicator_sqft_lot15', 'year', 'quarter', 'weekday'],
      dtype='object')
```

Some ML algorithms can be used with categorical variables, but others require the features to be in numerical format. It might make sense to convert them anyway, if for example, we plan to compare different algorithms against each other. One simple option is to just assign a random integer for each value. This is called label encoding or ordinal encoding. However, this approach induces an arbitrary ordering for the values, which might confuse some algorithms (although 1 < 2, it doesn't make sense to state that 'cat' < 'dog'). If a variable is ordinal (like the variable `condition` in our case) and we __control the order__, this might be OK. 

Complete the `condition_encoder` function that uses label encoding to convert the values in column `condition` in a given DataFrame to numerical ones. Use scaling poor=1, tolerable=2, satisfactory=3, good=4, and excellent=5. 

Hint: The [pandas.Series.map](https://pandas.pydata.org/docs/dev/reference/api/pandas.Series.map.html#pandas.Series.map) function may be useful. 

In [46]:
def condition_encoder(df: pd.DataFrame) -> None:
    """
    Encodes conditions to numerical range 1-5
    Args:
        df (DataFrame): A pandas DataFrame holding the features
    """

    ### START CODE HERE
    conversion_dict = {'poor':1, 'tolerable':2, 'satisfactory':3, 'good':4, 'excellent':5}
    df['condition'] = df['condition'].map(conversion_dict)
    ### END CODE HERE



Run the cell below to apply the encoder for the column `condition` in both `X_train` and `X_test`. You should see the following statistics:

| | X_train | X_test |
| :-: | :-: | :-: |
| mean | 3.362665 | 3.296886 |
| std | 0.526798 | 0.622107 |
| min | 3 | 1 |
| max | 5 | 5| 

In [47]:
condition_encoder(X_train)
condition_encoder(X_test)
print("Statistics for X_train:\n", X_train['condition'].describe())
print("\nStatistics for X_test:\n", X_test['condition'].describe())

Statistics for X_train:
 count    11046.000000
mean         3.362665
std          0.526798
min          3.000000
25%          3.000000
50%          3.000000
75%          4.000000
max          5.000000
Name: condition, dtype: float64

Statistics for X_test:
 count    2762.000000
mean        3.296886
std         0.622107
min         1.000000
25%         3.000000
50%         3.000000
75%         4.000000
max         5.000000
Name: condition, dtype: float64


In many cases, using label encoding is not advisable, because it implies that the values are ordered. For instance, take a look at the variable `postcode`, which although in numerical format, is actually a categorical variable. If we used them as is, a linear regressor model would have to assume that either a bigger postcode implies a bigger price or a smaller price. If (and when) this is not the case, the (assumably) valuable information in this column would be rendered pretty much useless.

Another option for converting categorical variables into numeric format is to use one-hot-encoding, where a binary feature is created for each unique value of the variable. This often leads to sparse feature matrices, when the cardinality of the categorical variable is high. For example, the `area` column has 131 unique values, so using one-hot-encoding would require us to present this variable using 130 additional features.  

One popular option is to use [target encoding](https://scikit-learn.org/stable/modules/preprocessing.html#target-encoder). The simplest way to use target encoding in a regression setting is to calculate the mean of target values and use that to encode categorical features. This might make sense, since for example, it's highly likely that the price of a house correlates with the area on which it is built. To reduce overfitting, the calculated means of categories can be (Laplace) smoothed with the overall mean of the target. Thus, the encoding $e_c$ for category $c$ becomes
$$
e_c = \lambda_c \cdot \bar{y}_c + (1-\lambda_c) \cdot \bar{y}, 
$$
where $\bar{y}_c$ is the category mean and $\bar{y}$ is the overall mean of the target and $\lambda \in [0,1]$ is a parameter controlling the smoothing amount. 

Complete the `target_encode` function which applies target encoding for a set of columns in a given DataFrame using the scikit-learn implementation. It should behave as follows: 
- If the target column is specified, the methods `fit` and `transform` should both be used (separately, don't use `fit_transform` to disable cross-fitting). 
- If the target column is not specified (for test data), only the method `transform` should be used with a pre-trained encoder passed as an argument.
- It replaces the values in the original column with the encoded ones.

NOTE: _The implementation of "fit_transform" in scikit-learn uses cross-fitting internally to avoid over-fitting. One thing needs to be considered at the point: target encoding may also need to be applied to production data (i.e., the real-world data the model used to generate predictions). However, targets are not available during inference, the encodings derived using cross-fitting can not be used. We skip cross-fitting entirely for simplicity in this exercise and rely solely on smoothing._

In [102]:
def target_encode(df: pd.DataFrame, columns: List[str], target: Optional[pd.Series]=None, encoder: Optional[TargetEncoder]=None) -> TargetEncoder:
    """
    Encodes postcode and area to numerical format using a target encoder
    Args:
        df (DataFrame): A pandas DataFrame holding the features
        columns (list of strings): Names of the categorical columns to be encoded
        target (Series|None): A pandas Series with the target values. This is required only when fitting the encoder.
        encoder(TargetEncoder|None): An already fitted encoder. This is required when we want to apply an encoder, which has already been fitted during training.
    Returns:
        encoder(TargetEncoder): The fitted TargetEncoder. Either a new fitted one or the one passed as an argument.
    """
    if encoder is None:
        encoder = TargetEncoder(target_type='continuous', smooth='auto', random_state=RANDOM_SEED)

    ### START CODE HERE
    data_to_encode = df.loc[:,columns]
    if target is not None:
        #encoder = TargetEncoder(smooth='auto')
        encoder = encoder.fit(data_to_encode, target)
    encoded_df = pd.DataFrame(data=encoder.transform(data_to_encode), columns=columns)
    df.loc[:, columns] = encoded_df
    ### END CODE HERE
    
    df.loc[:, columns].astype('float64')
    return encoder

    

Use the function to target-encode columns `postcode` and `area` in both `X_train` and `X_test` by running the cell below. You should see the following statistics:

| | X_train postcode | X_test postcode | X_train area | X_test area |
| :-: | :-: | :-: | :-: | :-: |
| mean | 5.919559e+05 | 5.991515e+05 | 5.897071e+05 | 5.963635e+05 |
| std | 1.519866e+05 | 1.602493e+05 | 1.458486e+05 | 1.534161e+05 |
| min | 1.506176e+05 | 1.506176e+05 | 1.378392e+05 | 1.619446e+05 |
| max | 1.275222e+06 | 1.275222e+06| 1.298436e+06 | 1.298436e+05 |

In [103]:
t_encoder = target_encode(X_train, ['postcode', 'area'], target=y_train)
_ = target_encode(X_test, ['postcode', 'area'], encoder=t_encoder)

print("Statistics for X_train postcode:\n", X_train['postcode'].describe())
print("\nStatistics for X_test postcode:\n", X_test['postcode'].describe())
print("Statistics for X_train area:\n", X_train['area'].describe())
print("\nStatistics for X_test area:\n", X_test['area'].describe())

Statistics for X_train postcode:
 count    1.104600e+04
mean     5.919559e+05
std      1.519866e+05
min      1.506176e+05
25%      4.819832e+05
50%      5.560479e+05
75%      6.038467e+05
max      1.275222e+06
Name: postcode, dtype: float64

Statistics for X_test postcode:
 count    2.762000e+03
mean     5.991515e+05
std      1.602493e+05
min      1.506176e+05
25%      4.819832e+05
50%      5.560479e+05
75%      6.995927e+05
max      1.275222e+06
Name: postcode, dtype: float64
Statistics for X_train area:
 count    1.104600e+04
mean     5.897071e+05
std      1.458486e+05
min      1.378392e+05
25%      4.812238e+05
50%      5.568487e+05
75%      6.575301e+05
max      1.298436e+06
Name: area, dtype: float64

Statistics for X_test area:
 count    2.762000e+03
mean     5.963635e+05
std      1.534131e+05
min      1.619446e+05
25%      4.857015e+05
50%      5.568487e+05
75%      6.742563e+05
max      1.298436e+06
Name: area, dtype: float64


We have now constructed a somewhat feasible feature extraction pipeline and derived a set of features. In real life, this process would be much more iterative (and time consuming) with a lot of experiments on different ways on how to build the features. We bypassed a lot of topics, such as scaling and normalization of numerical data, which are important parts for the optimal use of many ML algorithms. As already mentioned, this topic is dealt with more thoroughly on the [I2ML]((https://studies.helsinki.fi/courses/course-unit/hy-CU-118207827-2021-08-01)) course.

The Internet is also full of (opinionated) blog posts on feature engineering. These posts make a good source for new ideas on how to do things, but remember to always reserve a hint of scepticism and try to think things through yourself. As a final note, the scikit-learn library offers nice tools for building complex feature engineering pipelines, which we omitted in these exercises due to lack of time. Feel free to [check it out](https://scikit-learn.org/stable/modules/compose.html).

### 4c) Persisting derived features and encoders

To conclude this exercise, let's store the derived features and the fitted target encoder for future use. In a more demanding setup, we would use a feature store (like [Feast](https://feast.dev/)) for this, but in this exercise, we use just a dedicated folder in the local file system, namely the folder `feature_store`. 

Create the following functions:
1. `store_features`, which takes a set of features and stores them in the `feature_store` folder with a user-specified name in `.parquet` format.
1. `store_targets`, which takes a set of targets and stores them in the `feature_store` folder with a user-specified name in `.csv` format. To avoid potential issues in the remaining assignments, don't include index when saving targets into a CSV file. 
1. `store_encoder`, which takes a fitted `TargetEncoder` and stores it in the subfolder `encoders` inside the `feature_store` folder with a user specified name in `.pkl` (pickle) format.

Hint: Pickle files are written in binary format. Use mode 'wb' when writing the pickle file.

NOTE: _Pickle is the default format for persisting fitted models in scikit-learn. However, it is not secure. [Only unpickle data you trust](https://docs.python.org/3/library/pickle.html)._ 

In [106]:
def store_features(X: pd.DataFrame, feature_file_path: str) -> None:    
    """
    Stores a set of features to a specified location
    Args:
        X (DataFrame): A pandas DataFrame holding the features
        feature_file_path (Path): Path for the stored features, e.g., feature_store/housing_train_X.parquet
    """
    
    ### START CODE HERE
    X.to_parquet(feature_file_path)
    ### END CODE HERE



def store_targets(y: pd.Series, target_file_path: Path) -> None:
    """
    Stores a set of features to a specified location
    Args:
        y (Series): A pandas Series holding the target values
        target_file_path (Path): Path for the stored targets, e.g., feature_store/housing_train_y.csv
    """

    ### START CODE HERE
    y.to_csv(target_file_path, index=False)
    ### END CODE HERE



def store_encoder(encoder: TargetEncoder, encoder_file_path: Path) -> None:
    """
    Stores a targetEncoder to a specified location
    Args:
        encoder (TargetEncoder): A fitted scikit-learn TargetEncoder object
        encoder_file_path (Path): Path of the stored target encoder.
    """

    ### START CODE HERE
    pickle.dump(encoder, open(encoder_file_path, 'wb'))
    ### END CODE HERE

Let's again test the above functions. Run the cell below to see if a set of new files appears in your `feature_store` folder.

In [107]:
feature_store_path = WORKING_DIR / "feature_store"
housing_train_x_path = feature_store_path / "housing_train_X.parquet"
housing_test_x_path = feature_store_path / "housing_test_X.parquet"
housing_train_y_path = feature_store_path / "housing_train_y.csv"
housing_test_y_path =  feature_store_path / "housing_test_y.csv"
housing_target_encoder_path = feature_store_path / "encoders" / "housing_target_encoder.pkl"

store_features(X_train, housing_train_x_path)
store_features(X_test, housing_test_x_path)
store_targets(y_train, housing_train_y_path)
store_targets(y_test, housing_test_y_path)
store_encoder(t_encoder, housing_target_encoder_path)

## Assignment 5: Feature importance and feature selection (2 points)
Feature importance indicates how useful a feature is for a model to predict the target. Feature selection refers to the process of reducing the number of input variables in model training to save computational cost and even boost model performance. These topics are dealt in more detail in the [I2ML]((https://studies.helsinki.fi/courses/course-unit/hy-CU-118207827-2021-08-01)) course, so we'll just gently touch on them here.

### 5a) Basic trial with random forest regressor

Let's start by doing a simple comparison to see if the features we have prepared through the feature engineering steps are any good. In this exercise, we'll use [RandomForestRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html) from the scikit-learn library to train two models, each use different data. One model is trained with the "fully engineered" feature set we derived through the feature engineering steps in Assignment 4 and the other is trained with a "minimally engineered" dataset. We use [coefficient of determination $R^2$](https://en.wikipedia.org/wiki/Coefficient_of_determination) as our metric to evaluate the models (higher is better) and perform no hyperparameter optimization (which will be addressed in next week's exercises).

Do the following:
1. Prepare the minimally engineered data. 
    1. Split the training and testing DataFrames `train_df` and `test_df` provided below into features and targets. Specifically, `X_train_basic` contains the un-engineered features for the training dataset, `y_train_basic` contains the target values for the training dataset, `X_test_basic` contains the un-engineered features for the testing dataset, and `y_test_basic` contains the target values for the testing dataset. You can use the function `separate_X_and_y` implemented in Assignment 4.
    1. Remove all non-numerical columns (including the column `date`) and all columns with missing values from both `X_train_basic` and `X_test_basic`, since these might cause problems for the regressor models.
1. Read the fully engineered feature set and the target set for both training and test data from your `feature_store` directory. Similarly, `X_train` contains the ( fully engineered) features for the training dataset, `y_train` contains the corresponding target values for the training dataset, `X_test` contains the (fully engineered) features for the testing dataset, and `y_test` contains the corresponding target values for the testing dataset.
1. Train a RandomForestRegressor model `housing_rfr_basic` with `X_train_basic` and `y_train_basic` and evaluate the performance with `X_test_basic` and `y_test_basic` using $R^2$ as score.
1. Train another model `housing_rfr` using data `X_train` and `y_train` from Exercise 4. Evaluate the models performance with `X_test` and `y_test` using $R^2$ as score.

(The expected values are $R^2_{\text{basic}} = 0.89742$ and $R^2_{\text{engineered}} = 0.93228$).

In [ ]:
# Prepare data
train_path = WORKING_DIR / "data" / "reference" / "train"
test_path = WORKING_DIR / "data" / "reference" / "test"

# DataFrames of training and test data (not yet feature-engineered)
train_df = data_extraction(train_path, ['excellent', 'good', 'satisfactory'])
test_df = data_extraction(test_path, ['poor', 'tolerable', 'satisfactory', 'good', 'excellent'])

In [ ]:
# Random forest regressor that will be trained using the minimally engineered data
housing_rfr_basic = RandomForestRegressor(n_jobs=-1, random_state=RANDOM_SEED)

# Random forest regressor that will be trained using the feature-engineered data
housing_rfr = RandomForestRegressor(n_jobs=-1, random_state=RANDOM_SEED)

# Names of columns that should not be included when using data without feature engineering to train a model
# These columns are categorical or include missing values
removed_columns = ['postcode', 'area', 'date', 'condition', 'sqft_living15', 'sqft_lot15']

# R2 score of the model trained using minimally engineered data
r2_score_basic = 0

# R2 score of the model trained using the feature-engineered data
r2_score = 0

### START CODE HERE
X_train_basic, y_train_basic = None
X_test_basic, y_test_basic = None

X_train = None
y_train = None
X_test = None
y_test = None

### END CODE HERE

print("The R^2 score for the test set using only basic (minimally-engineered) features is", r2_score_basic)    
print("The R^2 score for the test set using the engineered features is", r2_score)    

### 5b) The most important features
The random forest regressor implementation in sklearn allows access to the attribute [feature_importances_](https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_importances.html#feature-importance-based-on-mean-decrease-in-impurity), which scores the features used in training using the (normalized) total reduction of error brought by that feature (higher values imply more important feature).

Your task is to draw a plot to show the importance of features used to train the `housing_rfr` model, the plot should be similar to the one in the [sklearn documentation](https://scikit-learn.org/stable/auto_examples/ensemble/plot_forest_importances.html#feature-importance-based-on-mean-decrease-in-impurity).

In [ ]:
### BEGIN SOLUTION

### END SOLUTION

You should see the feature `sqft_living` is significantly more important than other features. 

### 5c) Select only the most important features
In this exercise, you'll gain an intuition of how feature selection can affect model performance. 

Do the following:
1. Create a loop where a new RandomForestRegressor model is fitted on each iteration beginning with just the most important feature from `X_train` (the fully engineered feature set). Then, increase the number of important features by one in each iteration so the model is trained with all the features at the end. In other words, you need to train a model using the $i+1$ most important features of `X_train` in $i^{th}$ iteration (suppose i starts with zero). You also need to evaluate the model performance using $R^2$ score on each iteration.
1. Record the number of features and the $R^2$ score in each iteration. 
1. Plot the $R^2$ scores from each iteration as a function of the number of features.

Hint: You might find [numpy.argsort](https://numpy.org/doc/1.23/reference/generated/numpy.argsort.html) useful.

In [ ]:
# The number of important features that yield the best R2 score
optimal_number_of_features = 0

# The best r2 score
best_r2_score = 0

# The list of features that produce the best r2 score
best_features = []

### START CODE HERE

### END CODE HERE

print(f"Using the {optimal_number_of_features} most important features resulted in the best R^2 score {best_r2_score}.")
unimportant_features = [x for x in X_train.columns if x not in best_features]
print(f"These important features are: {best_features}, while the unimportant features are {unimportant_features}")

Expected output:
```text
Using the 16 most important features resulted in the best R^2 score 0.9344191835567803.
These important features are: Index(['sqft_living', 'postcode', 'area', 'waterfront', 'sqft_basement',
       'sqft_living15', 'yr_built', 'view', 'condition', 'bathrooms',
       'sqft_lot15', 'sqft_lot', 'distance', 'grade', 'year', 'bedrooms'],
      dtype='object'), while the unimportant features are ['floors', 'similarity_scores', 'missingindicator_sqft_living15', 'missingindicator_sqft_lot15', 'quarter', 'weekday']
```

We now know which features can result in the best $R^2$ score, let's implement the `drop_unimportant_features` that drop those unimportant columns.

In [ ]:
def drop_unimportant_features(df: pd.DataFrame):
    """
    df (DataFrame): DataFrame from which the unimportant features should be dropped.
    """
    unimportant_features = ['floors', 'similarity_scores', 'missingindicator_sqft_living15', 'missingindicator_sqft_lot15', 'quarter', 'weekday']
    ### START CODE HERE
    
    ### END CODE HERE

In [ ]:
# Drop unimportant features
X_train = pd.read_parquet(housing_train_x_path)
X_test = pd.read_parquet(housing_test_x_path)
drop_unimportant_features(X_train)
drop_unimportant_features(X_test)

print(X_train.columns)

Expected output:
```text
Index(['yr_built', 'bedrooms', 'postcode', 'area', 'bathrooms', 'condition',
       'grade', 'sqft_living', 'sqft_lot', 'sqft_basement', 'sqft_living15',
       'sqft_lot15', 'waterfront', 'view', 'distance', 'year'],
      dtype='object')
```

In [ ]:
# Save only important features
store_features(X_train, feature_store_path / "housing_train_X_important.parquet")
store_features(X_test, feature_store_path / "housing_test_X_important.parquet")

To conclude Assignments 4 and 5, let's again wrap everything inside one function that will be responsible for feature engineering. The function takes the data that has been preprocessed and validated, extracts all of the features, and stores them into the feature store. It's worth noting that this function will not only be applied to training data but also to test data and production data. As a result, the following three scenarios need to be considered:
|dataset type|target included|fit a new target encoder|load a fitted target encoder|
|------------|---------------|------------------------|----------------------------|
|training|x|x||
|test|x||x|
|production|||x|

In [ ]:
def feature_engineering_pipeline(df: pd.DataFrame, 
                                 feature_store_path: Path,
                                 feature_file_name: str, 
                                 encoder_file_name: str, 
                                 target_file_name: Optional[str]=None, 
                                 fit_encoder: bool=False, 
                                 targets_included: bool=True) -> None:
    """
    Converts a given (merged) housing data DataFrame into features and targets, performs feature engineering, and 
    stores the features along with possible targets and a fitted encoder
    Args:
        df (DataFrame): A pandas DataFrame holding all of the housing data, or just the features (see targets_included)
        feature_store_path (Path): Path of the feature store
        feature_file_name (str): Filename for the stored features.
        encoder_file_name (str): Filename for the stored encoder.
        target_file_name (str|None): Filename for the stored targets.
        fit_encoder (bool): Whether a new target encoder should be fitted. If False, uses a previously stored encoder
        targets_included (bool):  If True, df has all of the housing data including targets. If False, df has only the features.
    """
    if targets_included:
        X, y = separate_X_and_y(df, target='price')   
    else:
        if fit_encoder:
            raise ValueError("Target encoder can not be trained without targets.")
        X = df.copy()

    impute_missing(X)
    datetime_decomposer(X, dt_column_name='date')
    condition_encoder(X)
    drop_unimportant_features(X)

    feature_file_path = feature_store_path / feature_file_name
    target_file_path = feature_store_path / target_file_name
    encoder_file_path = feature_store_path / "encoders" / encoder_file_name

    if fit_encoder:
        t_encoder = target_encode(X, columns=['postcode', 'area'], target=y)
        store_features(X, feature_file_path)
        store_targets(y, target_file_path)
        store_encoder(t_encoder, encoder_file_path)
        
    else:
        with open(encoder_file_path, 'rb') as enc_f:
            t_encoder = pickle.load(enc_f)
        target_encode(X, columns=['postcode', 'area'], encoder=t_encoder)
        store_features(X, feature_file_path)    
        if targets_included:
            store_targets(y, target_file_path)

## Assignment 6: One script to rule them all (2 points)

The jupyter notebook we have played with this far is a good tool for exploratory and iterative development, but contains all sort of redundant code, which is no longer needed once we want to push the finalized version into production. As a final wrap-up exercise, we'll collect all of the functionalities created in this assignment to a single streamlined `etl.py` module. The file `etl.py` already exists as a template with all the necessary imports and global variables. 

### 6a) Let's begin with copying the functions...

Copy the following functions to `etl.py`: file_reader, dataframe_merger, drop_futile_columns, correct_distance_unit, string_transformer, typo_fixer, data_extraction, batch_creator, create_checkpoint, separate_X_and_y, impute_missing, datetime_decomposer, condition_encoder, target_encode, store_features, store_targets, store_encoder, drop_unimportant_features, feature_engineering_pipeline. 

There are many ways you can achieve this. Here are three:
1. Just copy-paste each function (carefully).
1. Use jupyter cell magic. You can add the line `%%writefile -a etl.py` at the beginning of each cell that contains a function. Run each of these cells exactly once. This will append the cell content inside the `etl.py` file. The cells should not contain code outside the function definitions. Once you've done, remove the cell magic commands.
1. Use `jupyter nbconvert` function. For example, running the following command will copy all your Python code (while ignoring Markdown cells) to a Python script that has the same name as the notebook. You can then clean the exported Python script and copy needed code such as the code skeleton of the `etl` function from `etl.py`. (Remember to rename your script to `etl.py` once you're done, otherwise you'll encounter issues when running some of the following code cells.)
```bash
jupyter nbconvert --to python --PythonExporter.exclude_markdown=True --no-prompt <jupyter-botebook-name>.ipynb 
```
### 6b) ...and create a wrapper

Create a wrapper function `etl` that loads, merges, cleans, and validates the specified data, extract features, and save the features (and possibly targets) in the feature store. Specifically, the function needs to perform as follows:
- It first runs the `data_extraction` function, which will result in a merged, cleaned DataFrame. you can use the hard-coded `correct_condition_values=['poor', 'tolerable', 'satisfactory', 'good', 'excellent']` in the `data_extraction` function.
- Then it uses Great Expectations Checkpoint to validate the DataFrame produced by the `data_extraction` function (The `batch_creator` and `create_checkpoint` functions are useful here). You can assume that the Great Expectations Context and the Expectation Suite are already configured. The GX Context is available via `context = gx.get_context(context_root_dir=gx_context_root_dir)` where `gx_context_root_dir` is the directory that contains all your Great Expectations config, which is the `gx` directory in our case. 
- If some validations fails, a warning should be printed. The warning can be as simple as a sentence like "Some GX validations failed". Hint: Please refer to [here](https://docs.greatexpectations.io/docs/reference/api/checkpoint/types/checkpoint_result/checkpointresult_class/) on how to check whether a validation succeeds or not. 
- Finally, no matter whether or not the DataFrame passes all Expectations, the `etl` function should feature-engineer the DataFrame using the `feature_engineering_pipeline` function. 

You can see a detailed description of the parameters of the `etl` function in [etl.py](./etl.py). 

Remember to include the completed `etl.py` in your submission.

Now, let's import the `etl` function and use it to process data from 2016.

In [ ]:
from etl import etl

params = {
        "path": WORKING_DIR / "data" / "2016",
        "gx_context_root_dir": WORKING_DIR / "gx",
        "gx_datasource_name": "housing_datasource", 
        "gx_checkpoint_name": "checkpoint_2016",
        "gx_expectation_suite_name": "housing_expectation_suite",
        "gx_run_name": "run_2016",
        "feature_store_path": WORKING_DIR / "feature_store",
        "feature_file_name": "2016_test_X.parquet",
        "encoder_file_name": "housing_target_encoder.pkl", 
        "target_file_name": "2016_test_y.csv", 
        "fit_encoder": False,
        "targets_included": True
    }
etl(**params)

You should see a warning telling that some Expectations failed. 

**Please attach the failed Expectations to your PDF file**, e.g., 

<img src="./images/failed-expectation-example.png" />

(Please note that the provided example doesn't show the failed validations for data from 2016. It just illustrates which parts of the Checkpoint result file should be included in the submitted PDF.)

Let's train a model using only the important features of the training set.

In [ ]:
train_X_important = pd.read_parquet(feature_store_path / "housing_train_X_important.parquet")
train_y = pd.read_csv(feature_store_path / "housing_train_y.csv").values.ravel()
housing_rfr2 = RandomForestRegressor(n_jobs=-1, random_state=RANDOM_SEED)
housing_rfr2.fit(train_X_important, train_y)

Then, evaluate the `housing_rfr2` model using data from 2016. 

In [ ]:
X_test_2016 = pd.read_parquet(WORKING_DIR/"feature_store"/"2016_test_X.parquet")
y_test_2016 = pd.read_csv(WORKING_DIR/"feature_store"/"2016_test_y.csv")
r2_score = housing_rfr2.score(X_test_2016, y_test_2016)
print(f"Model's R2 score using data from 2016 as test data: {r2_score}")

Model's R2 score using data from 2016 as test data should be 0.92359378. 